Add takeover based HUP script

kb2ma opened the pull request:
This script is equivalent to `upgrade-2.x.sh`, but for takeover based HUP. This script is derived from `upgrade-2.x.sh`, but removes unnecessary functionality. It also aims to be a model for a script implementation but avoiding gratuitous changes to the `upgrade-2.x.sh` original.

I am looking for higher level comments on the approach at this point. I have added some questions in the code. Don't mind the individual commits. I will squash them later.

kb2ma commented on upgrade-takeover.sh:
```diff
@@ -0,0 +1,460 @@
+    # Don't use --fail; we want the status code
+    curl_no_fail="curl --silent --retry 10 --location --compressed"
+
+    status_code=$(\
```
Worthwhile to capture status code?

alexgg replied:
Probably useful in the log to know the cause for failure.

kb2ma commented on upgrade-takeover.sh:
```diff
@@ -0,0 +1,460 @@
+takeover_path="/mnt/data/takeover"
+mkdir -p "${takeover_path}"
+
+# Download target hostOS image as required
```
Is it feasible to avoid re-downloading the same OS image in case of running again after failure?

kb2ma commented on upgrade-takeover.sh:
```diff
@@ -0,0 +1,460 @@
+esac
+
+# Ensure takeover environment is prepared
+takeover_path="/mnt/data/takeover"
```
Is it reasonable to copy the takeover binary and OS image to this directory. OS image is approximately 250 MB.

kb2ma replied:
I see that I can get the content-length header from the API call. Is the best approach to compare against the output from `df` for the partition mounted on `/mnt/data` -- or is there an easier way to grab that available size?

alexgg replied:
Something like `df --output=avail /mnt/data/ | tail -n 1` should work well enough

alexgg commented on upgrade-takeover.sh:
```diff
@@ -0,0 +1,460 @@
+mkdir -p "${takeover_path}"
+
+# Download target hostOS image as required
+if [ ! -f "${takeover_path}/balenaos.img.gz" ]; then
```
How do you know this is the same as the requested version? I would honestly just cleanup on exit and re-download.

alexgg commented on upgrade-takeover.sh:
```diff
@@ -0,0 +1,460 @@
+
+# Download takeover binary as required
+if [ ! -f "${takeover_path}/takeover" ]; then
+    download_takeover_binary "v0.9.0-dev.1"
```
Why not always download the latest version? Would we ever release a new version that we don't want to use?

kb2ma replied:
I agree. This version is a pre-release just for testing until we have a binary that is reasonably useful, so it can be called 'latest'.

alexgg commented on upgrade-takeover.sh:
```diff
@@ -0,0 +1,460 @@
+if [ ! -f "${takeover_path}/takeover" ]; then
+    download_takeover_binary "v0.9.0-dev.1"
+else
+    log "Takeover binary available; download not required"
```
I would clean-up on exit and always re-download.

alexgg commented on upgrade-takeover.sh:
```diff
@@ -0,0 +1,460 @@
+# We do *not* expect takeover to return on success; it flashes the device with
+# the new balenaOS.
+res=$(./takeover -i balenaos.img.gz \
+   --no-ack --no-nwmgr-check --no-os-check --no-vpn-check \
```
Why no-vpn-check? Isn't it better to default to only update if the device has a fair chance of being remotely accessible?

kb2ma replied:
@rahul-thakoor , any concerns with including the VPN check? Should we set `--check-timeout`? What is your experience?

kb2ma replied:
@acostach ^^

alexgg commented on upgrade-takeover.sh:
```diff
@@ -0,0 +1,460 @@
+# the new balenaOS.
+res=$(./takeover -i balenaos.img.gz \
+   --no-ack --no-nwmgr-check --no-os-check --no-vpn-check \
+   --log-level debug -l /dev/sda1 --s2-log-level debug)
```
Is `-l /dev/sda1`  the target device? If so it should not be hard coded as different devices will have different storage. It should be deduced from the running system I reckon, like inspecting the host OS to see what parent device is the active partition mounted on.

kb2ma replied:
Yes, that is for logging to an external device. That was just there until we incorporated the on-device logging. See balena-os/takeover#100

kb2ma commented on upgrade-takeover.sh:
```diff
@@ -0,0 +1,460 @@
+    rm "${takeover_path}/takeover.tar.gz"
+}
+
+# Download target balenaOS image; expects a public image
```
I understand the API download endpoint _does_ support private images.  I am confirming with those in the know since I don't have such a device. I'll update the comment.

kb2ma commented on upgrade-takeover.sh:
```diff
@@ -0,0 +1,460 @@
+# Remove and log fsck recovery files. Since takeover will reflash the disk, it's
+# not strictly necessary. However, it may indicate issues with the storage medium,
+# so good to know.
+function remove_rec_files() {
```
How does `fsck` help in this case if takeover will overwrite all of the partitions?

alexgg replied:
I was following up on the comment above - `fsck` or similar would show whether the media is healthy before copying and launching the takeover tool. How takeover would fail from a damaged storage is difficult to predict - probably failing to write and handling the error, but it's probably an untested path.

However, if it fails to write because of damaged sectors, it could have already overwritten parts so the device will be bricked. Bailing out if fsck errors is a way to improve reliability.

kb2ma replied:
Sounds fine to me. So to be clear, you're saying use `fsck -n`, capture stdout+stderr, and failure is defined as any non-zero result. Log the output on failure. Use `lsblk` to find the device name of the `/mnt/data` partition?

Let me know of any better ways; not my expertise.

alexgg replied:
Yeah, something like:
```
fsck.ext4 -n /dev/sda1 > dev/null
if [ $? -ge 4 ]; then
    exit 1
fi
```

kb2ma replied:
Done; to be clear, only checking data partition.

kb2ma commented:
@alexgg Fixed up from your review and rebased. All changes are in the final commit. In addition to your review, I did these things:

* Renamed "takeover_path" to "work_dir". It's more descriptive, especially since adding the cleanup function
* Added some options to the takeover binary to accommodate how it works now
